services/federator/test/integration/Test/Federator/InwardSpec.hs

```diff
@@ -73,24 +76,60 @@ spec env =
         _ <- putHandle brig (userId user) hdl
 
         let expectedProfile = (publicProfile user UserLegalHoldNoConsent) {profileHandle = Just (Handle hdl)}
+        bdy <- asInwardBodyUnsafe <$> inwardBrigCall "federation/get-user-by-handle" (encode hdl)
+        liftIO $ bdy `shouldBe` expectedProfile
 
-        Endpoint fedHost fedPort <- cfgFederatorExternal . view teTstOpts <$> ask
-        client <- createGrpcClient (grpcClientConfigSimple (Text.unpack fedHost) (fromIntegral fedPort) False)
-        c <- case client of
-          Left err -> liftIO $ assertFailure (show err)
-          Right cli -> pure cli
-        let brigCall = Request Brig "federation/get-user-by-handle" (LBS.toStrict (encode hdl)) "foo.example.com"
-        res <- liftIO $ gRpcCall @'MsgProtoBuf @Inward @"Inward" @"call" c brigCall
-
-        liftIO $ case res of
-          -- TODO: When brig returns 404, this does not correctly parse as
-          -- InwardResponseErr, this needs to be looked into.
-          GRpcOk (InwardResponseBody bdy) ->
-            eitherDecodeStrict bdy `shouldBe` Right expectedProfile
-          GRpcOk (InwardResponseErr err) -> assertFailure $ "Unexpected error response: " <> show err
-          x -> assertFailure $ "GRpc call failed unexpectedly: " <> show x
-
--- All the code below is copied from brig-integrration tests
+    it "should give InvalidEndpoint on a 404 'no-endpoint'" $
+      runTestFederator env $ do
+        err <- asInwardErrorUnsafe <$> inwardBrigCall "federation/this-endpoint-does-not-exist" (encode Aeson.emptyObject)
+        expectErr IInvalidEndpoint err
+
+-- Utility functions
+--
+expectErr :: InwardErrorType -> InwardError -> TestFederator IO ()
+expectErr expectedType err = do
+  case inwardErrorType err of
+    t | t == expectedType -> pure ()
+    _ -> liftIO $ assertFailure $ "expected type '" <> show expectedType <> "' but got " <> show err
+
+asInwardBodyUnsafe :: (HasCallStack, Typeable a, FromJSON a) => GRpcReply InwardResponse -> a
+asInwardBodyUnsafe = either err id . asInwardBodyEither
+  where
+    err parserErr = error . unwords $ ["asInwardBodyUnsafe:"] <> [parserErr]
```

Comment:
Why not assert that the response is a body, instead of using `error`? I found that using `error` in tests can sometimes cause some failures not to be reported, because the test code can run into the error call before it has had a chance to process and print the failures.

Reply:
I tried both using `assertFailure` and `error`. The error messages are different, it worked fine locally for me, and I liked the display of using `error` 10% more. We use `error` quite a bit in other integration tests by using `responseJsonUnsafe` and companions.
I switched this PR to make use of assertFailure in https://github.com/wireapp/wire-server/pull/1637/commits/aacb5182eb3cc3cb2fdf304a1ec5ae438d8f6943 (if this is what you had in mind?), but in general I'm curious to see where you've seen this problem that failures are not reported?

Reply:
I'm not sure where exactly (and it's gone now), but it was indeed a piece of test code using `responseJsonUnsafe`. It looked like:
```haskell
doStuff !!! do
  const 200 === statusCode
  blah === responseJsonUnsafe
```
and I only saw a parse failure error in a situation where the status code was *not* 200. This was confusing, because the assertion that parses the body comes after the one about the status code, but you only get to see the former.

Reply:
Ah I see, okay that's a good enough argument to not use error I suppose.